phpBB/phpbb/notification/type/mention.php

```diff
@@ -0,0 +1,404 @@
+	{
+		if (count($users_to_notify) > 0)
+		{
+			$sql_ary = "SELECT users.user_id, users.username_clean FROM ". USERS_TABLE . " as users, " . USER_NOTIFICATIONS_TABLE . " as notif WHERE " . $db->sql_in_set('username_clean', $users_to_notify) . " AND users.user_id = notif.user_id AND notif.item_type = 'notification.type.mention' " ;
+			$result = $db->sql_query($sql_ary);
+			$user_list = array();
+			while ($row = $db->sql_fetchrow($result))
+			{
+				$user_list[$row["username_clean"]] = $row["user_id"];
+			}
+			$db->sql_freeresult($result);
+			return $user_list;
+		}
+		return array();
+	}
+
+	/**
+	* Get the type and method of notification corresponding the user
+	*
+	* @return \Array Array containing the notification type object and notification
+	* method object.
+	*/
+	public function get_notification_type_and_method($db, $user_list, $notif_manager)
+	{
+		$notif_list = array();
+		if (count($user_list) > 0)
+		{
+			$integer_ids = array_map('intval', $user_list);
+			$sql_query = "SELECT user_id, item_type, method FROM " . USER_NOTIFICATIONS_TABLE . " where " . $db->sql_in_set("user_id", $integer_ids) . " AND item_type = 'notification.type.mention'";
+			$result = $db->sql_query($sql_query);
+			while ($row = $db->sql_fetchrow($result))
+			{
+				// $temp_notif_list = array();
+				if (is_array($notif_list[$row["user_id"]]))
+				{
+					$notif_list[$row["user_id"]][] = $row["method"];
+				}
+				else
+				{
+					$notif_list[$row["user_id"]] = array();
+					$notif_list[$row["user_id"]][] = $row["method"];
+				}
+			}
+			$db->sql_freeresult($result);
+		}
+		return $notif_list;
+	}
+
+	/**
+	* Get the url to this item
+	*
+	* @return string URL
+	*/
+	public function get_url()
+	{
+		return append_sid($this->phpbb_root_path . 'viewtopic.' . $this->php_ext, "p={$this->item_id}#p{$this->item_id}");
+	}
+
+	/**
+	* {inheritDoc}
+	*/
+	public function get_redirect_url()
+	{
+		return append_sid($this->phpbb_root_path . 'viewtopic.' . $this->php_ext, "t={$this->item_parent_id}&amp;view=unread#unread");
+	}
+
+	/**
+	* Get email template
+	*
+	* @return string|bool
+	*/
+	public function get_email_template()
+	{
+		return 'mention_notify';
+	}
+
+	/**
+	* Get email template variables
+	*
+	* @return array
+	*/
+	public function get_email_template_variables()
+	{
+		if ($this->get_data('post_username'))
+		{
+			$username = $this->get_data('post_username');
+		}
+		else
+		{
+			$username = $this->user_loader->get_username($this->get_data('poster_id'), 'username');
+		}
+
+		return array(
+			'AUTHOR_NAME'               => htmlspecialchars_decode($username),
+			'POST_SUBJECT'              => htmlspecialchars_decode(censor_text($this->get_data('post_subject'))),
+			'TOPIC_TITLE'               => htmlspecialchars_decode(censor_text($this->get_data('topic_title'))),
+			'U_VIEW_POST'               => generate_board_url() . "/viewtopic.{$this->php_ext}?p={$this->item_id}#p{$this->item_id}",
```

Comment:
I traced the codeflow and I guess its right, if I have not missed out something. It starts with posting.php where send_notification is called , inside send_notification() the data array is sent to add_notification_for_users() -> on line 321 inside the manager class get_item_id() method is called and the post_id is set as $this->item_id inside the get_item_id() method.